x/wasm/keeper/addresses.go

```diff
@@ -0,0 +1,58 @@
+package keeper
+
+import (
+	"encoding/binary"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/address"
+
+	"github.com/CosmWasm/wasmd/x/wasm/types"
+)
+
+// AddressGenerator abstract address generator to be used for a single contract address
+type AddressGenerator func(ctx sdk.Context, codeID uint64, checksum []byte) sdk.AccAddress
+
+// ClassicAddressGenerator generates a contract address using codeID and instanceID sequence
+func (k Keeper) ClassicAddressGenerator() AddressGenerator {
+	return func(ctx sdk.Context, codeID uint64, _ []byte) sdk.AccAddress {
+		instanceID := k.autoIncrementID(ctx, types.KeyLastInstanceID)
+		return BuildContractAddressClassic(codeID, instanceID)
+	}
+}
+
+// PredicableAddressGenerator generates a predictable contract address
+func PredicableAddressGenerator(creator sdk.AccAddress, salt []byte, msg []byte, fixMsg bool) AddressGenerator {
+	return func(ctx sdk.Context, _ uint64, checksum []byte) sdk.AccAddress {
+		if !fixMsg { // clear msg to not be included in the address generation
+			msg = []byte{}
+		}
+		return BuildContractAddressPredictable(checksum, creator, salt, msg)
+	}
+}
+
+// BuildContractAddressClassic builds an sdk account address for a contract.
+func BuildContractAddressClassic(codeID, instanceID uint64) sdk.AccAddress {
+	contractID := make([]byte, 16)
+	binary.BigEndian.PutUint64(contractID[:8], codeID)
+	binary.BigEndian.PutUint64(contractID[8:], instanceID)
+	return address.Module(types.ModuleName, contractID)[:types.ContractAddrLen]
+}
+
+// BuildContractAddressPredictable generates a contract address for the wasm module with len = types.ContractAddrLen using the
+// Cosmos SDK address.Module function. Similar to BuildContractAddressPredictable1 but including the initMsg.
+// Internally a key is built containing:
+// (len(checksum) | checksum | len(sender_address) | sender_address | len(salt) | salt| len(initMsg) | initMsg).
+//
+// All method parameter values must be valid and not nil.
+func BuildContractAddressPredictable(checksum []byte, creator sdk.AccAddress, salt, initMsg []byte) sdk.AccAddress {
+	checksum = address.MustLengthPrefix(checksum)
+	creator = address.MustLengthPrefix(creator)
+	salt = address.MustLengthPrefix(salt)
+	initMsg = address.MustLengthPrefix(initMsg)
```

Comment:
Here we need a different solution since 1 byte lengths is not enough

Reply:
you are right. 255 may not be enough!

Reply:
I have pushed a change. The slice can not be > max int  https://stackoverflow.com/questions/27647737/maximum-length-of-a-slice-in-go